allensdk/brain_observatory/filtered_running_speed/__main__.py

```diff
@@ -0,0 +1,954 @@
+        angular_change,
+        frame_times
+    )  # speed in radians/s
+
+    linear_speed = deg_to_dist(angular_speed)
+
+    # Artifact correction to speed data
+    wrap_corrected_linear_speed = _clip_speed_wraps(
+        linear_speed, frame_times, np.concatenate([pos_wraps, neg_wraps]),
+        t_span=0.25)
+
+    outlier_corrected_linear_speed = _zscore_threshold_1d(
+        wrap_corrected_linear_speed, threshold=zscore_threshold)
+
+    # Final filtering (optional) for smoothing out the speed data
+    if lowpass:
+        polynomial_b, polynomial_a = signal.butter(
+            FILTER_ORDER,
+            Wn=CRITICAL_FREQUENCY,
+            fs=DEFAULT_SAMPLING_FREQUENCY,
+            btype="lowpass"
+        )
+
+        outlier_corrected_linear_speed = signal.filtfilt(
+            polynomial_b,
+            polynomial_a,
+            np.nan_to_num(outlier_corrected_linear_speed)
+        )
+
+    dx_rad = degrees_to_radians(dx_raw)
+
+    frame_indexes = list(range(start_index, end_index))
+
+    df = pd.DataFrame(
+        {
+            "velocity": outlier_corrected_linear_speed[:len(frame_times)],
+            "net_rotation": dx_rad[:len(frame_times)],
+            "frame_indexes": frame_indexes,
+            "frame_times": frame_times
+        }
+    )
+
+    # due to an acquisition bug (the buffer of raw orientations may be updated
+    # more slowly than it is read, leading to a 0 value for the change in
+    # orientation over an interval) there may be exact zeros in the velocity.
+    df = df[~(np.isclose(df["net_rotation"], 0.0))]
+
+    return df
+
+
+def match_timestamps(num_raw_timestamps, signal, label):
+    """
+    Sometimes the timestamp has one fewer value than
+    the signal does, if this is the case, we can remove the extra value
+    from the end of the signal
+
+    Argument defaults and implementation suggestion via @dougo
+
+    Parameters
+    ----------
+    num_raw_timestamps: list
+        A list of timestamps
+    signal: list
+        A list of signal values
+    label: string
+        Name of the signal
+
+    Returns
+    -------
+    list
+        An adjusted signal to match the timestamp's length
+    """
+
+    if num_raw_timestamps == (len(signal) - 1):
+        signal = signal[0:len(signal) - 1]
+
+    if num_raw_timestamps != len(signal):
+        raise ValueError(
+            f"found {num_raw_timestamps} rising edges on the vsync line, "
+            f"but found {len(signal)} samples for  {label}"
+        )
+
+    return signal
+
+
+def extract_dx_info(
+    frame_times,
+    start_index,
+    end_index,
+    pkl_path
+):
+    """
+    Extract all of the running speed data
+
+    Parameters
+    ----------
+    frame_times: numpy.ndarray
+        list of the vsync times
+    start_index: int
+        Index to the first frame of the stimulus
+    end_index: int
+       Index to the last frame of the stimulus
+    pkl_path: string
+        Path to the stimulus pickle file
```

Comment:
Just checking: do we need to specify which stimulus pickle file this is, or is it any of the three?

Reply:
This has been refactored.